api/src/pcapi/core/finance/models.py

```diff
@@ -271,3 +271,90 @@ class Invoice(Model):
     # See the note about `amount` at the beginning of this module.
     amount = sqla.Column(sqla.Integer, nullable=False)
     url = sqla.Column(sqla.Text, nullable=False, unique=True)
+
+
+class ReferenceScheme(Model):
+    """This table holds the next reference number (and other related
+    parameters) for our invoices and possibly other entities.
+
+    The goal is to have a way to generate consecutive reference
+    numbers without gaps. We cannot use a serial sequence, because
+    transaction rollbacks do not "decrease" the sequence and that
+    would cause a gap. Instead, we store the next reference in a row
+    of this table and update it only within the same transaction that
+    sets that reference on the related entity (e.g. ``invoice``).
+
+    Usage:
+
+        scheme = models.ReferenceScheme.get_and_lock(name="invoice.reference", year=year)
+        invoice.reference = scheme.formatted_reference
+        scheme.increment_after_use()
+        db.session.add(invoice)
+        db.session.add(scheme)
+        db.session.commit()
+
+    It is very important that there is a single transaction that
+    encloses all calls. If a transaction commits the use of the
+    reference model and another transaction commits the
+    incrementation, there is a risk that the latter fails and that the
+    """
+
+    id = sqla.Column(sqla.BigInteger, primary_key=True, autoincrement=True)
+    # known names are:
+    #   - "invoice.reference"
+    name = sqla.Column(sqla.Text, nullable=False)
+    prefix = sqla.Column(sqla.Text, nullable=False)
+    year = sqla.Column(sqla.Integer)
+    nextNumber = sqla.Column(sqla.Integer, default=1)
+    numberPadding = sqla.Column(sqla.Integer, default=7)
+
+    __table_args__ = (
+        sqla.UniqueConstraint(
+            "name",
+            "year",
+            name="unique_name_year",
+        ),
+        sqla.UniqueConstraint(
+            "prefix",
+            "year",
+            name="unique_prefix_year",
+        ),
+    )
+
+    @classmethod
+    def get_and_lock(cls, name, year=None):
+        # Here we wait for the lock. If another transaction has a
+        # lock, that's fine: we'll wait for that other transaction to
+        # commit and release the lock, so that we get the latest,
+        # updated reference.
+        return (
+            cls.query.populate_existing()
+            .with_for_update(nowait=False)
+            .filter_by(
+                name=name,
+                year=year,
+            )
+            .one()
+        )
+
+    def increment_after_use(self) -> None:
```

Comment:
Pourrait-on faire un try/except autour du `FOR UPDATE NOWAIT` ci-dessous, et lever une erreur spécifique le cas échéant: `ReferenceIncrementError` par exemple ?

Reply:
Bonne idée, c'est fait. Merci !